File: artifacts/original/ProjectOne/src/contacts/ContactService.java
/*
 * Jeremy Cameron
 * CS 320
 * ContactService class which is used for creating and managing contacts.
 */

package contacts;

import java.util.HashMap;
import java.util.Map;

/*
 * ContactService class used for managing contacts.
 * It provides methods for adding, deleting, retrieving, and updating contacts. 
 */
public class ContactService {
	
	/*
	 * Creates a hash map to store contacts, using the contact ID as the key
	 * and the contact object as the value. Used to efficiently manage contacts by their unique IDs.
	 */
	private final Map<String, Contact> contacts = new HashMap<>();
	
	/*
	 * Adds a contact to the service.
	 * Throws an IllegalArgumentException if a duplicate contact ID.
	 */
	public void addContact(Contact contact) {
		if (contacts.containsKey(contact.getId())) {
			throw new IllegalArgumentException("Contact ID already exists");
		}
		contacts.put(contact.getId(), contact);
	}
	
	/*
	 * Deletes a contact from the service.
	 * Throws an IllegalArgumentException if contact ID does not exist.
	 */
	public void deleteContact(String contactId) {
		if (!contacts.containsKey(contactId)) {
			throw new IllegalArgumentException("Contact ID does not exist");
		}
		contacts.remove(contactId);
	}
	
	/*
	 * Updates the contact's first name.
	 * Throws an IllegalArgumentException if contact ID does not exist.
	 */
	public void updateFirstName(String contactId, String firstName) {
		if (!contacts.containsKey(contactId)) {
			throw new IllegalArgumentException("Contact ID does not exist");
		}
		contacts.get(contactId).setFirstName(firstName);
	}
	
	/*
	 * Updates the contact's last name.
	 * Throws an IllegalArgumentException if contact ID does not exist.
	 */
	public void updateLastName(String contactId, String lastName) {
		if (!contacts.containsKey(contactId)) {
			throw new IllegalArgumentException("Contact ID does not exist");
		}
		contacts.get(contactId).setLastName(lastName);
	}
	
	/*
	 * Updates the contact's phone number.
	 * Throws an IllegalArgumentException if contact ID does not exist.
	 */
	public void updatePhoneNumber(String contactId, String phoneNumber) {
		if (!contacts.containsKey(contactId)) {
			throw new IllegalArgumentException("Contact ID does not exist");
		}
		contacts.get(contactId).setPhoneNumber(phoneNumber);
	}
	
	/*
	 * Updates the contact's address.
	 * Throws an IllegalArgumentException if contact ID does not exist.
	 */
	public void updateAddress(String contactId, String address) {
		if (!contacts.containsKey(contactId)) {
			throw new IllegalArgumentException("Contact ID does not exist");
		}
		contacts.get(contactId).setAddress(address);
	}
	
	/*
	 * Retrieves a contact by its ID.
	 * Throws an IllegalArgumentException if contact ID does not exist.
	 */
	public Contact getContact(String id) {
		if (!contacts.containsKey(id)) {
			throw new IllegalArgumentException("Contact ID does not exist");
		}
		return contacts.get(id);
	}
}

File: enhancement_2/CS499_Project/src/services/contact-service.js
// src/services/contact-service.js
import Contact from '../domain/contact.js';
import DuplicateIdError from '../errors/error-duplicate-id.js';
import { ensureIdExists } from '../utils/validators.js';
import { Trie } from '../utils/trie.js';

export default class ContactService {
  // Creates a new private Map to store contacts by ID
  #contacts = new Map(); 
  // Creates new private Tries for storing first and last names
  #trieFirstName = new Trie();
  #trieLastName = new Trie();

  // Method to add a new contact
  addContact(data) {
    // If ID is provided, checks if it already exists.
    const idToCheck = data.id ?? null;
    if (idToCheck && this.#contacts.has(idToCheck)) {
      throw new DuplicateIdError(`Contact ID already exists: ${idToCheck}`);
    }
    // Creates a new contact and checks if the generated ID already exists
    const contact = new Contact(data);
    if (this.#contacts.has(contact.id)) {
      throw new DuplicateIdError(`Contact ID already exists: ${contact.id}`);
    }
    // Stores the new contact in the Map and returns it
    this.#contacts.set(contact.id, contact);
    // Stores the first and last names inside the Trie
    this.#trieFirstName.insert(contact.firstName, contact.id);
    this.#trieLastName.insert(contact.lastName, contact.id);
    return contact;
  }

  // Method to delete a contact
  deleteContact(id) {
    // Retrieves the contact
    const contact = this.getContact(id);
    // Deletes the first and last names stored in the Trie
    this.#trieFirstName.delete(contact.firstName, id);
    this.#trieLastName.delete(contact.lastName, id);
    // Deletes the contact ID from the Map
    this.#contacts.delete(id);
  }

  // Method to update a contact's first name
  updateFirstName(id, newFirstName) {
    const contact = this.getContact(id);
    // Checks if old first name matches the new first name
    if (contact.firstName != newFirstName) {
      // Deletes the old first name from the Trie
      this.#trieFirstName.delete(contact.firstName, id);
      // Updates new first name into the Trie
      contact.setFirstName(newFirstName);
      this.#trieFirstName.insert(newFirstName, id);
    }
    return contact;
  }

  // Method to update a contact's last name
  updateLastName(id, newLastName) {
    // Retrieves the contact
    const contact = this.getContact(id);
    // Checks if old last name matches the new last name
    if (contact.lastName != newLastName) {
      // Deletes the old last name from the Trie
      this.#trieLastName.delete(contact.lastName, id);
      // Updates new last name into the Trie
      contact.setLastName(newLastName);
      this.#trieLastName.insert(newLastName, id);
    }
    return contact;
  }

  // Method to update a contact's phone number
  updatePhoneNumber(id, phoneNumber) {
    const contact = this.getContact(id);
    contact.setPhoneNumber(phoneNumber);
    return contact;
  }

  // Method to update a contact's address
  updateAddress(id, address) {
    const contact = this.getContact(id);
    contact.setAddress(address);
    return contact;
  }

  // Method to search contacts by prefix for first name
  searchFirstNamePrefix(prefix) {
    const ids = this.#trieFirstName.searchPrefix(prefix);
    return Array.from(ids).map(id => this.#contacts.get(id).toJSON());
  }

  // Method to search contacts by prefix for last name
  searchLastNamePrefix(prefix) {
    const ids = this.#trieLastName.searchPrefix(prefix);
    return Array.from(ids).map(id => this.#contacts.get(id).toJSON());
  }

  // Method to search contacts by prefix for both first and last name
  searchFullNamePrefix(prefixFirstName, prefixLastName) {
    const idsFirstName = this.#trieFirstName.searchPrefix(prefixFirstName);
    const idsLastName = this.#trieLastName.searchPrefix(prefixLastName);
    const nameIntersect = new Set([...idsFirstName].filter(id => idsLastName.has(id)));
    return Array.from(nameIntersect).map(id => this.#contacts.get(id).toJSON());
  }

  // Method to retrieve a contact
  getContact(id) {
    // Verifies the ID exists before trying to retrieve
    ensureIdExists(this.#contacts, id, 'Contact');
    return this.#contacts.get(id);
  }

  // Method to retrieve all contacts as an array JSON objects
  getAllContacts() {
    return Array.from(this.#contacts.values()).map(c => c.toJSON());
  }
}